41-vaccine-temp-monitor/firmware/arduino/src/main.cpp

```diff
@@ -0,0 +1,528 @@
+#ifdef ARDUINO_ARCH_ESP32
+#include <NotecardAuxiliaryWiFi.h>
+#endif // ARDUINO_ARCH_ESP32
+
+// Uncomment this line and replace com.your-company:your-product-name with your
+// ProductUID.
+// #define PRODUCT_UID "com.your-company:your-product-name"
+
+#ifndef PRODUCT_UID
+#define PRODUCT_UID "" // "com.my-company.my-name:my-project"
+#pragma message "PRODUCT_UID is not defined in this example. Please ensure your Notecard has a product identifier set before running this example or define it in code here. More details at https://bit.ly/product-uid"
+#endif
+#define usbSerial Serial
+#define txRxPinsSerial Serial1
+
+#define CRT_MAX_TEMP_C 30
+#define CRT_SAFE_TEMP_C 25
+#define DEFAULT_SAMPLE_INTERVAL_S 1800
+#define RETRY_S 5
+
+M2M_LM75A lm75a;
+Notecard notecard;
+NotecardEnvVarManager *envVarManager = NULL;
+#ifdef ARDUINO_ARCH_ESP32
+using namespace blues;
+NotecardAuxiliaryWiFi aux_wifi(notecard);
+#endif
+
+volatile int alarm_hysteresis = CRT_SAFE_TEMP_C;
+volatile int alarm_temperature = CRT_MAX_TEMP_C;
+volatile size_t sample_interval_s = DEFAULT_SAMPLE_INTERVAL_S;
+
+void logNoteF(const char *format_, ...)
+{
+    char log[256];
+    va_list args;
+    va_start(args, format_);
+    vsnprintf(log, sizeof(log), format_, args);
+
+    if (J *cmd = notecard.newCommand("hub.log"))
+    {
+        JAddStringToObject(cmd, "text", log);
+        if (!notecard.sendRequest(cmd))
+        {
+            JDelete(cmd);
+        }
+    }
+
+    va_end(args);
+}
+
+int acquireGPSLocation(size_t timeout_s_ = 95)
+{
+    int result;
+    const char *gps_mode;
+    size_t gps_time_s;
+
+    // Get Current Configuration
+    if (J *rsp = notecard.requestAndResponse(notecard.newRequest("card.location.mode")))
+    {
+        gps_mode = JGetString(rsp, "mode");
+        NoteDeleteResponse(rsp);
+    }
+    else
+    {
+        result = -1; // Unable to communicate with Notecard
+        return result;
+    }
+
+    // Halt periodic tracking, before active sync is performed
+    if (J *cmd = notecard.newCommand("card.location.mode"))
+    {
+        JAddStringToObject(cmd, "mode", "off");
+        notecard.sendRequest(cmd);
+    }
+
+    // Gather timestamp of previous location information
+    if (J *rsp = notecard.requestAndResponse(notecard.newRequest("card.location")))
+    {
+        gps_time_s = JGetInt(rsp, "time");
+        logNoteF("INFO: Previous GPS time (seconds): %u", gps_time_s);
+        NoteDeleteResponse(rsp);
+    }
+    else
+    {
+        result = -1; // Unable to communicate with Notecard
+        return result;
+    }
+
+    // Activate GPS
+    if (J *cmd = notecard.newCommand("card.location.mode"))
+    {
+        JAddStringToObject(cmd, "mode", "continuous");
+        notecard.sendRequest(cmd);
+    }
+
+    // Block while resolving GPS location
+    for (const size_t start_ms = ::millis();;)
+    {
+        if (timeout_s_)
+        {
+            if (::millis() >= (start_ms + (timeout_s_ * 1000)))

```

Comment:
I think this has the millis rollover problem. https://arduino.stackexchange.com/questions/12587/how-can-i-handle-the-millis-rollover